File: solutions/python/stack/backspace-string-compare.py
# LeetCode 844 — Backspace String Compare (Easy)
# Category: Stack · Approach: Stack
# Time: O(n + m) | Space: O(n + m)
# Source: https://leetcode.com/problems/backspace-string-compare/

def backspace_compare(s: str, t: str) -> bool:
    def build(text: str) -> str:
        stack: list[str] = []
        for c in text:
            if c == "#":
                if stack:
                    stack.pop()
            else:
                stack.append(c)
        return "".join(stack)

    return build(s) == build(t)

File: solutions/typescript/stack/backspace-string-compare.ts
// LeetCode 844 — Backspace String Compare (Easy)
// Category: Stack · Approach: Stack
// Time: O(n + m) | Space: O(n + m)
// Source: https://leetcode.com/problems/backspace-string-compare/

function backspaceCompare(s: string, t: string): boolean {
  const build = (text: string): string => {
    const stack: string[] = [];
    for (const c of text) {
      if (c === "#") stack.pop();
      else stack.push(c);
    }
    return stack.join("");
  };
  return build(s) === build(t);
}
